embassy-imxrt/src/gpio.rs

```diff
@@ -0,0 +1,1056 @@
+//! GPIO
+
+use core::convert::Infallible;
+use core::future::Future;
+use core::marker::PhantomData;
+use core::pin::Pin as FuturePin;
+use core::task::{Context, Poll};
+
+use embassy_hal_internal::interrupt::InterruptExt;
+use embassy_sync::waitqueue::AtomicWaker;
+use sealed::Sealed;
+
+use crate::clocks::enable_and_reset;
+use crate::iopctl::IopctlPin;
+pub use crate::iopctl::{AnyPin, DriveMode, DriveStrength, Function, Inverter, Pull, SlewRate};
+use crate::{interrupt, peripherals, Peri, PeripheralType};
+
+// This should be unique per IMXRT package
+const PORT_COUNT: usize = 8;
+
+/// Digital input or output level.
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Level {
+    /// Logic Low
+    Low,
+    /// Logic High
+    High,
+}
+
+impl From<bool> for Level {
+    fn from(val: bool) -> Self {
+        match val {
+            true => Self::High,
+            false => Self::Low,
+        }
+    }
+}
+
+impl From<Level> for bool {
+    fn from(level: Level) -> bool {
+        match level {
+            Level::Low => false,
+            Level::High => true,
+        }
+    }
+}
+
+/// Interrupt trigger levels.
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum InterruptType {
+    /// Trigger on level.
+    Level,
+    /// Trigger on edge.
+    Edge,
+}
+
+#[cfg(feature = "rt")]
+#[interrupt]
+#[allow(non_snake_case)]
+fn GPIO_INTA() {
+    irq_handler(&GPIO_WAKERS);
+}
+
+#[cfg(feature = "rt")]
+struct BitIter(u32);
+
+#[cfg(feature = "rt")]
+impl Iterator for BitIter {
+    type Item = u32;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.0.trailing_zeros() {
+            32 => None,
+            b => {
+                self.0 &= !(1 << b);
+                Some(b)
+            }
+        }
+    }
+}
+
+#[cfg(feature = "rt")]
+fn irq_handler(port_wakers: &[Option<&PortWaker>]) {
+    let reg = unsafe { crate::pac::Gpio::steal() };
+
+    for (port, port_waker) in port_wakers.iter().enumerate() {
+        if port_waker.is_none() {
+            continue;
+        }
+
+        let stat = reg.intstata(port).read().bits();
+        for pin in BitIter(stat) {
+            // Clear the interrupt from this pin
+            reg.intstata(port).write(|w| unsafe { w.status().bits(1 << pin) });
+            // Disable interrupt from this pin
+            reg.intena(port)
+                .modify(|r, w| unsafe { w.int_en().bits(r.int_en().bits() & !(1 << pin)) });
```

Comment:
Is it possible that between clearing the interrupt status and disabling the interupt for the pin, another edge/level event happens which would set the status register to a high value?

It does appear that the interrupt is cleared in the constructor of `InputFuture`, so that would prevent a new instance of InputFuture seeing this event.

Reply:
What is the concern here? That the same event could retrigger due to e.g. bounces? The future is consumed and creating a new one clears any potential left-over events. I don't such opportunity. If I missed the point, let me know.

Reply:
Probably not a concern since the InputFuture clears it anyways and interrupt is disabled.